lib/Types.cpp

```diff
@@ -120,4 +121,59 @@ void PostScaling::parse(const nlohmann::json &composition)
         }
     }
 }
+
+Resolution::Resolution()
+    : numerator(1)
+    , denominator(1)
+{
+}
+
+Resolution::Resolution(uint64_t numerator, uint64_t denominator)
+    : numerator(numerator)
+    , denominator(denominator)
+{
+}
+
+void Resolution::compose(nlohmann::json &composition) const
+{
+    composition[META_RESOLUTION][META_NUMERATOR] = numerator;
+    composition[META_RESOLUTION][META_DENOMINATOR] = denominator;
+}
+
+int Resolution::parse(const nlohmann::json &composition, LogCallback logCallback)
+{
+    int result = 0;
+    uint64_t newNumerator = numerator;
+    uint64_t newDenominator = denominator;
+
+    auto resolutionIter = composition.find(META_RESOLUTION);
+    if (resolutionIter!=composition.cend()) {
+        auto numeratorIter = resolutionIter->find(META_NUMERATOR);
+        if(numeratorIter!=resolutionIter->cend()) {
+            newNumerator = *numeratorIter;
+            if (newNumerator==0) {
+                STREAMING_PROTOCOL_LOG_E("\tResolution numerator may not be 0!");
+                return -1;
+            }
+            result = 1;
+        }
+        auto denominatorIter = resolutionIter->find(META_DENOMINATOR);
+        if(denominatorIter!=resolutionIter->cend()) {
+            newDenominator = *denominatorIter;
+            if (newDenominator==0) {
+                STREAMING_PROTOCOL_LOG_E("\tResolution denominator may not be 0!");
+                return -1;
+            }
+            result = 1;
+        }
+    }
+    numerator = newNumerator;
+    denominator = newDenominator;
+    return result;
```

Comment:
In general, it seems more logical to return 0 (instead of 1) on success and a negative value on failure.

Reply:
I agree, the reason for this hack is that there are three states.

-error (-1)
-parsed successfully (1)
-no error but nothing to parse (0)

The caller should act only in the second case (1)-